text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,271 @@
+
+The approach that we propose for adoption is #3. Specifically:
+
+- We detect a mismatch
+- We patch "easily recoverable" mismatches in the DOM, e.g. text-node mismatches
+- For mismatches that are not easily recoverable
+  + We determine the closest LWC ancestor to that mismatch (anywhere from the root component in a SPA to a leaf component for a small widget).
+  + We invalidate the ancestor's subtree, replacing SSR-derived DOM with CSR DOM for that component subtree only.
+
+If we determine that this behavior has undesirable performance characteristics, we will pivot to approach #2.
+
+### Recovering from lifecycle errors
+
+It might happen that an error is thrown while invoking the `connectedCallback` or other lifecycle method. This situation could occur because:
+
+1. The userland lifecycle method has a bug.
+2. The state extracted from the SSR DOM is incorrect for some reason, leading to unexpected inputs and behavior in the lifecycle method.
+
+We will want to recover from #2. And because it won't be possible to differentiate #1 and #2 at runtime, a catch-all solution must be adopted.
+
+It is proposed that, should an error be thrown during `connectedCallback`, we should assume that an unexpected mismatch has occurred, throw away the SSR DOM, and replace it with CSR VDOM. The `connectedCallback` will then be invoked again (`connectedCallback` can fire more than once), this time for the CSR DOM.
+
+If an error occurs again, we can assume its proximate cause is in userland and not in LWC.
+
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(document.querySelector('#root'), MyApp, { some: props }); 
+```
+
+`hydrateElement` accepts three arguments:
+
+1. The already-existing DOM node where the root component should be attached.
+2. The LWC class intended for use as the root component.
+3. Any props that should be supplied to that root component as part of the initial client-side render.
+
+### Constructable stylesheets
+
+The introduction of constructable stylesheets to LWC presents a small wrinkle. Since support for this browser API is not yet ubiquitous, we cannot know during SSR whether it will be supported on the client. For that reason, constructable stylesheets are disabled during SSR.
+
+From there, we have two options if support for constructable stylesheets is detected during rehydration:
+
+1. Swap in constructable stylesheets for vanilla stylesheets.
+2. Permanenly disable constructable stylesheets for the lifetime of the `document`.

```

Comment:
```suggestion
2. Disable constructable stylesheets during hydration, for the lifetime of the component instance.
```

Explanation: The stylesheets are created, and cached in the VM the first time the VM renders ([code](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine-core/src/framework/template.ts#L144-L185)).